From c847879bc5558e7e141c417fa1650ab0609a7fcb Mon Sep 17 00:00:00 2001
From: "River, Li" <river.li@intel.com>
Date: Thu, 24 Dec 2015 18:25:39 +0800
Subject: [PATCH 06/10] FOR_UPSTREAM [VPG]: drm/i915: Add trace to gem request
 state.

This commit adds trace event to track gem request complete state.

Tracked-On: https://jira01.devtools.intel.com/browse/OAM-26200
Change-Id: I504dd306b9a4f273db4d4ad5ac8401182c6edeaa
Signed-off-by: River, Li <river.li@intel.com>
Reviewed-on: https://android.intel.com:443/477659
---
 drivers/gpu/drm/i915/i915_gem.c   |  2 ++
 drivers/gpu/drm/i915/i915_trace.h | 35 +++++++++++++++++++++++++++++++----
 2 files changed, 33 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index b6043c8..642f36c 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -3120,6 +3120,7 @@ void i915_gem_complete_requests_ring(struct intel_engine_cs *ring,
 	u32 seqno;
 
 	seqno = ring->get_seqno(ring, lazy_coherency);
+	trace_i915_gem_request_complete_begin(ring, seqno);
 	if (!seqno)
 		return;
 
@@ -3129,6 +3130,7 @@ void i915_gem_complete_requests_ring(struct intel_engine_cs *ring,
 
 	spin_lock_irqsave(&ring->reqlist_lock, flags);
 	list_for_each_entry(req, &ring->request_list, list) {
+		trace_i915_gem_request_complete_loop(req);
 		if (req->complete)
 			continue;
 
diff --git a/drivers/gpu/drm/i915/i915_trace.h b/drivers/gpu/drm/i915/i915_trace.h
index b6079eb..efa86d1 100644
--- a/drivers/gpu/drm/i915/i915_trace.h
+++ b/drivers/gpu/drm/i915/i915_trace.h
@@ -611,6 +611,7 @@ DECLARE_EVENT_CLASS(i915_gem_request,
 			     __field(u32, ring)
 			     __field(u32, uniq)
 			     __field(u32, seqno)
+			     __field(u32, completed)
 			     ),
 
 	    TP_fast_assign(
@@ -620,11 +621,11 @@ DECLARE_EVENT_CLASS(i915_gem_request,
 			   __entry->ring = ring->id;
 			   __entry->uniq = req ? req->uniq : 0;
 			   __entry->seqno = i915_gem_request_get_seqno(req);
+			   __entry->completed = i915_gem_request_completed(req);
 			   ),
-
-	    TP_printk("dev=%u, ring=%u, uniq=%u, seqno=%u",
-		      __entry->dev, __entry->ring, __entry->uniq,
-		      __entry->seqno)
+	    TP_printk("dev=%u, ring=%u, uniq=%u, seqno=%u, completed=%u",
+			__entry->dev, __entry->ring, __entry->uniq,
+			__entry->seqno, __entry->completed)
 );
 
 DEFINE_EVENT(i915_gem_request, i915_gem_request_add,
@@ -662,6 +663,11 @@ DEFINE_EVENT(i915_gem_request, i915_gem_request_complete,
 	    TP_ARGS(req)
 );
 
+DEFINE_EVENT(i915_gem_request, i915_gem_request_complete_loop,
+	    TP_PROTO(struct drm_i915_gem_request *req),
+	    TP_ARGS(req)
+);
+
 TRACE_EVENT(i915_gem_request_wait_begin,
 	    TP_PROTO(struct drm_i915_gem_request *req),
 	    TP_ARGS(req),
@@ -1321,6 +1327,27 @@ TRACE_EVENT(queue_retire_work,
 );
 /* queue_retire_work - end */
 
+TRACE_EVENT(i915_gem_request_complete_begin,
+	TP_PROTO(struct intel_engine_cs *ring, u32 seqno),
+
+	TP_ARGS(ring, seqno),
+
+	TP_STRUCT__entry(
+		__field(u32, ring)
+		__field(u32, seqno)
+		__field(u32, last_seqno)
+	),
+
+	TP_fast_assign(
+	    __entry->ring = ring->id;
+	    __entry->seqno = seqno;
+	    __entry->last_seqno = ring->last_read_seqno;
+	),
+
+	TP_printk("ring=%u,seqno=%d,last_seqno=%d",
+		__entry->ring, __entry->seqno, __entry->last_seqno)
+);
+
 #endif /* _I915_TRACE_H_ */
 
 /* This part must be outside protection */
-- 
1.9.1

